lib/internal/url.js

```diff
@@ -85,34 +85,36 @@ class TupleOrigin {
   }
 }
 
+// Resued by URL constructor and
+// URL#href setter.
+function parse(input, base) {
+  input = String(input);
+  const base_context = base ? base[context] : undefined;
+  this[context] = new StorageObject();
+  binding.parse(input.trim(), -1, base_context, undefined,
+    (flags, protocol, username, password,
+      host, port, path, query, fragment) => {
+      if (flags & binding.URL_FLAGS_FAILED)
+        throw new TypeError('Invalid URL');
+      this[context].flags = flags;
+      this[context].scheme = protocol;
+      this[context].username = username;
+      this[context].password = password;
+      this[context].port = port;
+      this[context].path = path;
+      this[context].query = query;
+      this[context].fragment = fragment;
+      this[context].host = host;
+      this[searchParams] = new URLSearchParams(query);
```

Comment:
Thanks for spotting this. Fixed and added tests. Could be in conflict with https://github.com/nodejs/node/pull/10399/files#diff-7fa9474fccb31c3ccef5d728fbd5248bL557 for reusing part of the URLSearchParams constructor though, should be easy to rebase.